File: src/main/java/uk/gov/justice/hmpps/offenderevents/services/MergeRecordDiscriminator.java
package uk.gov.justice.hmpps.offenderevents.services;

import com.microsoft.applicationinsights.TelemetryClient;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Component
@Slf4j
public class MergeRecordDiscriminator {
    private final PrisonApiService prisonApiService;

    private final TelemetryClient telemetryClient;

    public MergeRecordDiscriminator(PrisonApiService prisonApiService,
                                    TelemetryClient telemetryClient) {
        this.prisonApiService = prisonApiService;
        this.telemetryClient = telemetryClient;
    }

    public List<MergeOutcome> identifyMergedPrisoner(Long bookingId) {

        return prisonApiService.getPrisonerNumberForBookingId(bookingId)
            .map(prisonerNumber -> {
                log.debug("Check for merges for booking ID {}, prisoner Number {}", bookingId, prisonerNumber);
                final var mergedOffenders = prisonApiService.getIdentifiersByBookingId(bookingId);

                return mergedOffenders.stream()
                    .peek(mergedNumber -> {
                        final var trackingAttributes = Map.of(
                            "bookingId", bookingId.toString(),
                            "mergedFrom", mergedNumber.value(),
                            "mergedTo", prisonerNumber
                        );
                        telemetryClient.trackEvent("POEMergeEvent", trackingAttributes, null);
                        log.debug("Prisoner record merged {} --> {}", mergedNumber.value(), prisonerNumber);
                    })
                    .map(mergedNumber -> new MergeOutcome(mergedNumber.value(), prisonerNumber))
                    .collect(Collectors.toList());
            })
            .orElse(Collections.emptyList());
    }


    record MergeOutcome(String mergedNumber, String remainingNumber) {
    }

}

File: src/main/java/uk/gov/justice/hmpps/offenderevents/services/MergeRecordDiscriminator.kt
package uk.gov.justice.hmpps.offenderevents.services

import com.microsoft.applicationinsights.TelemetryClient
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class MergeRecordDiscriminator(
  private val prisonApiService: PrisonApiService,
  private val telemetryClient: TelemetryClient,
) {
  fun identifyMergedPrisoner(bookingId: Long): List<MergeOutcome> =
    prisonApiService.getPrisonerNumberForBookingId(bookingId)
      .map { prisonerNumber: String ->
        log.debug(
          "Check for merges for booking ID {}, prisoner Number {}",
          bookingId,
          prisonerNumber,
        )
        val mergedOffenders = prisonApiService.getIdentifiersByBookingId(bookingId)
        mergedOffenders.forEach { mergedNumber: BookingIdentifier ->
          val trackingAttributes = mapOf(
            "bookingId" to bookingId.toString(),
            "mergedFrom" to mergedNumber.value,
            "mergedTo" to prisonerNumber,
          )
          telemetryClient.trackEvent("POEMergeEvent", trackingAttributes, null)
          log.debug("Prisoner record merged {} --> {}", mergedNumber.value, prisonerNumber)
        }
        mergedOffenders
          .map { mergedNumber: BookingIdentifier -> MergeOutcome(mergedNumber.value, prisonerNumber) }
      }
      .orElse(emptyList())

  @JvmRecord
  data class MergeOutcome(val mergedNumber: String, val remainingNumber: String)

  private companion object {
    private val log = LoggerFactory.getLogger(this::class.java)
  }
}
